backend/onedrive/quickxorhash/quickxorhash.go

```diff
@@ -7,58 +7,88 @@
-// Which has the copyright
-
-// ------------------------------------------------------------------------------
-//  Copyright (c) 2016 Microsoft Corporation
+// This code was ported from a fast C-implementation from
+// https://github.com/namazso/QuickXorHash
+// which has licenced as BSD Zero Clause License
+// BSD Zero Clause License
 //
-//  Permission is hereby granted, free of charge, to any person obtaining a copy
-//  of this software and associated documentation files (the "Software"), to deal
-//  in the Software without restriction, including without limitation the rights
-//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-//  copies of the Software, and to permit persons to whom the Software is
-//  furnished to do so, subject to the following conditions:
+// Copyright (c) 2022 namazso <admin@namazso.eu>
 //
-//  The above copyright notice and this permission notice shall be included in
-//  all copies or substantial portions of the Software.
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted.
 //
-//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-//  THE SOFTWARE.
-// ------------------------------------------------------------------------------
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
+// OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
 
 import (
 	"hash"
+
+	"reflect"
+	"unsafe"
 )
 
 const (
 	// BlockSize is the preferred size for hashing
 	BlockSize = 64
 	// Size of the output checksum
-	Size           = 20
-	bitsInLastCell = 32
-	shift          = 11
-	widthInBits    = 8 * Size
-	dataSize       = (widthInBits-1)/64 + 1
+	Size        = 20
+	shift       = 11
+	widthInBits = 8 * Size
+	dataSize    = shift * widthInBits
 )
 
 type quickXorHash struct {
-	data        [dataSize]uint64
-	lengthSoFar uint64
-	shiftSoFar  int
+	data [dataSize]byte
+	size uint64
 }
 
 // New returns a new hash.Hash computing the quickXorHash checksum.
 func New() hash.Hash {
 	return &quickXorHash{}
 }
 
+func bytes2U64(v []byte) (lv []uint64) {
+	bh := (*reflect.SliceHeader)(unsafe.Pointer(&v))
```

Comment:
Can you remove these `unsafe` uses?

Reply:
unfortunately, that is the smallest solution to prevent usage of unsafe
conversion of []byte to []uint64 in place is required, otherwise compiler would calculate a byte per cycle or unnecessary memory copy in performance critical area

Reply:
In your original commit https://github.com/rclone/rclone/pull/6663/commits/f1897c2070f130d98f9fe3893f9f0b4f5501f099 you didn't have this use of unsafe. Can you put it back like that?

Reply:
According to the comments it'd reduce the speed from 8125 MB/s back to 3077 MB/s. I get that unsafe is not nice but is it really worth throwing away a 264% speed increase?

Reply:
There is another solution that uses package `binary`, that might take effect on removing `unsafe` and keeping acceptable performance (will still get worse), but the code will be more chaotic due to different solutions in different archs.  
The decline of 264% is really painful, so I agree with namazso 👍

Reply:
Could this use of unsafe disappear with go1.20?

Reply:
yes, it will move into runtime after go1.22 is released (rclone's go version policy)

Reply:
So could we make this with 3GB/s without `unsafe` and make a version with go1.20 which runs at 8GB/s also without `unsafe` in rclone code. I don't care if there is a use of `unsafe` in the crypto library as Google have better tools and more manpower than me to validate such uses.